### ISSUE TITLE
Mixin debugging support

### ISSUE THREAD

**Issue by Minecrell** [SpongePowered/Mixin#156]
Continued discussion from #155.

### The issue
Mixin classes are merged into the target classes at runtime. Because of that, debugging Mixins usually results in jumping around on invalid lines in the source of the target class instead of in the actual Mixin code.

The debugger resolves the source location using the class name and the `SourceFile` attribute of the Java class file, which is limited to a single source file. Consequently, debugging methods in the target class and a Mixin is not possible at the same time: the `SourceFile` can be only set to  the original source file or the source file of the Mixin.

### Using decompiled code
Up to now, the only solution for this problem was to use special support in the IDE (primarily IntelliJ IDEA) to debug using decompiled classes instead of the original sources. However, while you are able to see the classes exactly the way they are executed at runtime (including the invocation of injectors etc) it has several disadvantages:

- Requires manual setup to configure Mixin to dump the bytecode and add it to the module classpath
- Comments and the original code formatting is removed due to the decompilation
- Breakpoints can be only set after the Mixin has already been applied once. To add a breakpoint for a new Mixin you have to run the application once to dump the new bytecode and let the IDE refresh all classes until you can actually see your changes in the target class.

### Source debug extension
As already mentioned, Java supports only a single `SourceFile` attribute in the class file. This is not only a problem for Mixin, but also for other projects, primarily programming languages which are compiled to Java source code before getting compiled to the final bytecode.

A solution for this problem was added to Java as part of [JSR-045 (Debugging Support for Other Languages)](https://jcp.org/en/jsr/detail?id=45). It describes a new attribute in the Java class file format, the `SourceDebugExtension` attribute, together with a standardized Source Map Format (SMAP).

SMAP was primarily developed for languages which are first compiled to Java source files. It can map the final line numbers in the bytecode to other line numbers in one or multiple source files.

On the other hand, Mixin keeps all original line numbers from the Mixins that are applied to the target class. Therefore the line numbers in the final class are not unique, and cannot be mapped reliably to the original source.

### My suggested solution
A solution would be to assign unique line numbers to all merged Mixin methods. However, this would make reading stack traces even harder than it already is. Right now you can use the line number to find the Mixin which is causing an exception.

A much easier solution is to encode a custom format in the `SourceDebugExtension` which is specific to Mixin. The [Java class file format specification](https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.7.11) does not dictate a specific format for the attribute - it is possible to store any string in it. All tools I've checked verify if the string is a SMAP (using the `SMAP` prefix) before parsing it. It is also [exposed in raw form in Java's debugging interface (JDI)](https://docs.oracle.com/javase/8/docs/jdk/api/jpda/jdi/com/sun/jdi/ReferenceType.html#sourceDebugExtension--), so we can parse it in the debugger to provide Mixin debugging.

In general, the only thing the debugger needs to know is the Mixin the method in the target class was merged from. I've used a very simple format to encode the metadata in the attribute:

```
MIXIN
com.example.mixin.MixinExample run()V handler$onRun$zza000(Lorg/spongepowered/asm/mixin/injection/callback/CallbackInfo;)V
com.example.mixin.MixinExtra runExtra()V
```

The `MIXIN` prefix allows us to detect encoded data as Mixin source debug extension - this is very similar to the solution in the SMAP format which uses the `SMAP` indentifier.

### Example implementation
I've implemented debugging support for the IntelliJ IDEA plugin [MinecraftDev](https://github.com/minecraft-dev/MinecraftDev). It parses the additional metadata and uses it to jump to the correct source of the method. It will also redirect breakpoints to the target class so it will properly pause the execution at runtime.

See https://github.com/minecraft-dev/MinecraftDev/pull/115 for some screenshots of the Mixin debugging.

**Comment by Minecrell** [comment]
@Mumfrey 
> Why is this a PR and not an issue? Open an issue **first** to discuss things with me before making PRs out of the blue.

My intention for the PR was to provide a proof of concept for one of options we can choose. I didn't expect to get it merged as-is, instead I wanted to suggest one way to implement it. I don't mind moving the discussion to an issue.

> I don't see why you wouldn't use the SMAP format for this, there is already code to support it.

After looking at the debugger implementation of IntelliJ IDEA I came to my personal conclusion that implementing it with SMAP is actually much more work compared to my solution. The general JSR-045 support in IDEs appears to be close to non-existent, IntelliJ IDEA supports [JSR-045 specifically only for JSP](https://www.jetbrains.com/help/idea/2016.3/run-debug-configuration-jsr45-compatible-server.html) and only in the proprietary Ultimate edition. A quick search suggests the same for Eclipse, but I have neither checked its source code nor actively used it in the past few years.

All in all it appears to me that the SMAP format is barely supported in the Java community is - there is barely any documentation for it except for the original specification and its usages seems to be entirely centered around JSP / other web pages.

That means that I would actually have to copy everything IntelliJ IDEA has in place to resolve Java positions to add JSR-045 support, instead of being able to reuse certain parts of it like in my other solution. And that does not even include modifying Mixin to assign each merged method unique line numbers or finding a solution to make exceptions not entirely unreadable (because of random line numbers).

My conclusion to this was that it is simpler to use a custom format which only maps methods to classes instead of lines since that is the only thing we really need.

> Honestly I think if you want to do this, it would be more appropriate to support it by providing a mechanism to add extra plugins into Mixin, and then implement the IntelliJ plugin on that.

That is actually exactly how I started implementing this. I started with a simple tweaker class that registers a transformer after the Mixin transformer, executes similar code to what I had in the PR and finally writes the resulting map to the `SourceDebugExtension`. However, every time Mixin switched its phase it moved its transformer to the end of the chain so my transformer was invoked too early (I believe this is intended).

Consequently, my next step was to check the Mixin source for interfaces I could implement to generate the map after the Mixins have been applied. I was searching for some kind of plugin framework I could hook into. There is only one reason why I made the PR to Mixin instead of requesting some kind of plugin interface: the map isn't stricly specific to the IntelliJ IDEA plugin, an Eclipse plugin could use it the same way I do.

However, I don't know enough about Eclipse to be able to say if it would be really useful the same way it is for IntelliJ IDEA. If you prefer to have it separate from the Mixin core, implementing some kind of plugin interface from an extra tweaker would also work for the IntelliJ IDEA plugin.

> Calling the option `sourceDebugExtension` is misleading because it doesn't actually enable the standard SMAP format, it should be called something which references the fact that it only supports one particular, InteliJ plugin.

The reason for calling the option `sourceDebugExtension` is partly explained above, my intention was to develop a "standard Mixin source debug extension format" that could be also used by other IDE plugins, not only by MinecraftDev.

I don't think that the name implies that it would enable the SMAP format (then it would be called `sourceMap` or `smap`). The `SourceDebugExtension` is a general attribute for additional debugging metadata that can be used for SMAP, but also for any other format.

**Comment by Mumfrey** [comment]
First of all, thanks for making this issue. This is something I've played around a lot with in the past, especially with respect to trying to get SMAP to work and found that basically I can generate a perfectly valid SMAP and have it roundly ignored by basically everything, including the Java runtime, IntelliJ and Eclipse. So to your points.

> After looking at the debugger implementation of IntelliJ IDEA I came to my personal conclusion that implementing it with SMAP is actually much more work compared to my solution. The general JSR-045 support in IDEs appears to be close to non-existent, IntelliJ IDEA supports JSR-045 specifically only for JSP and only in the proprietary Ultimate edition. A quick search suggests the same for Eclipse, but I have neither checked its source code nor actively used it in the past few years.

What I would have hoped though, is that if there's support for SMAP in an IDE *at all* is that it might be possible to simply subclass/extend/fork/fix it work with Mixins perhaps. I feel like this should be a long term goal since the format itself is quite powerful. The approach I used was that each Mixin applied to a class got a base linenumber offset, eg. 1000 for the first, 2000 for the second, determined by stride, then when the Mixins were applied the linenumbers were shifted by that amount and written into the SMAP, with no "intermediate file" as is stipulated in the JSR45. However as I mentioned, I couldn't get any tool whatsoever to consume the data and the whole thing was binned.

> All in all it appears to me that the SMAP format is barely supported in the Java community is - there is barely any documentation for it except for the original specification and its usages seems to be entirely centered around JSP / other web pages.

There is very little documentation but the documentation which does exist is pretty comprehensive at least. The part about it being extremely JSP-oriented is very true though, I haven't found any consumers for it other than Tomcat.

> That means that I would actually have to copy everything IntelliJ IDEA has in place to resolve Java positions to add JSR-045 support, instead of being able to reuse certain parts of it like in my other solution. And that does not even include modifying Mixin to assign each merged method unique line numbers or finding a solution to make exceptions not entirely unreadable (because of random line numbers).

Right, I understand that the wiring might be a pain. The idea is a bit of a pipe dream but I want to support it eventually.
 
> My conclusion to this was that it is simpler to use a custom format which only maps methods to classes instead of lines since that is the only thing we really need.

I would agree with this, it's also the shortest route to market, which is good. Now the reason I mentioned "plugins" is there's a long-running local branch I have which is a solution to #108 (auditing), which basically introduces an observer model for all of Mixin's actions (basically taking the premise of that issue but extending it to all aspects of mixin activity). The observer receives a callback for each action in order to add it into the audit log. The native audit observer consumes the callbacks and uses them to build an XML document with an XSLT which allows it to be viewed in a browser.

What I would propose is that either:

* You consume the XML file thus generated - which has all the info you need: this is not that pretty though
* I change the consumer so that it can be pluggable (eg. via Java Service model) and then you provide your own consumer directly via the IntelliJ plugin, this would be quite powerful because you'd literally only need to put your plugin on the classpath for Mixin to find it, and also you could just sink the audit events you need. (In this case adding method to class)

This kills two birds with one stone as it's a reason to finish up the auditing stuff which has been pushed back in favour of other things, and also gives you an extensible point to hook your needed functionality onto. I could also give you access to the SMAP branch if you wanted to look into that, PM me on IRC about it though because it's too messy to post publically.

**Comment by Minecrell** [comment]
@Mumfrey That sounds good. I have sent you a PM on IRC for the SMAP branch - if you have a working solution that generates a SMAP I can play a bit with the JDI and IntelliJ API and see if I can get something working.

For the XML document: Would you generate one for every Mixin or one for all Mixin that have been applied? For the debugger we need to be quite careful because as soon as the class is loaded the data could be needed at any time, even if other classes haven't been loaded yet.

The same applies to the audit events: To make it work properly you would have to call an event after auditing is complete, together with the `ClassNode` so the plugin could generate the map based on the previously added methods. The JDI API is very limited, the attribute is more or less the only reliable way to have the data available as soon as it is needed.

**Comment by Mumfrey** [comment]
> @Mumfrey That sounds good. I have sent you a PM on IRC for the SMAP branch - if you have a working solution that generates a SMAP I can play a bit with the JDI and IntelliJ API and see if I can get something working.

Sure, though I dunno how much work is going to be needed to rebase it because the original was written in April '15 and I think I rebased it last some time earlier this year so it'll probably need some love before it works with the current build.

> For the XML document: Would you generate one for every Mixin or one for all Mixin that have been applied? For the debugger we need to be quite careful because as soon as the class is loaded the data could be needed at any time, even if other classes haven't been loaded yet.

It's one XML document but it gets serialised after every mixin application, same as the implementation report.

> The same applies to the audit events: To make it work properly you would have to call an event after auditing is complete, together with the ClassNode so the plugin could generate the map based on the previously added methods. The JDI API is very limited, the attribute is more or less the only reliable way to have the data available as soon as it is needed.

No, you misunderstand. Auditing happens in real time, so you would never see the ClassNode. You just get notifications *when* method `x` is added to class `y` by mixin `z`.

**Comment by Minecrell** [comment]
> No, you misunderstand. Auditing happens in real time, so you would never see the ClassNode. You just get notifications when method x is added to class y by mixin z.

So how could we use this in the IntelliJ IDEA plugin to have the data available when debugging?

**Comment by Mumfrey** [comment]
> So how could we use this in the IntelliJ IDEA plugin to have the data available when debugging?

Have your consumer pump the data into intellij using whatever method suits. Named pipe or socket would be the easiest but you could write it to disk for that matter.

**Comment by Minecrell** [comment]
> Sure, though I dunno how much work is going to be needed to rebase it because the original was written in April '15 and I think I rebased it last some time earlier this year so it'll probably need some love before it works with the current build.

I have rebased your changes on `master` and pushed the rebased commit to the `sourcemap-rebased` branch on your server. I've also done some tests with the IntelliJ IDEA debugger to see if we can get debugging working with the SMAP.

The reason why your implementation isn't working as expected seems to be related to the usage of the default `Java` stratum. The specification mentions:

> The final-source stratum (named "Java") is created automatically and should not have a stratum section.

My testing has confirmed this behaviour: JDI will automatically generate the line numbers and source name from the bytecode. Consequently, the data in your SMAP was simply silently ignored and replaced with the automatically generated mappings. A simple fix for this issue is to use a custom stratum name, e.g. `Mixin`.

I've also realized I was slightly wrong about the SMAP support: The implementation in JDI is actually implemented in a way that doesn't necessarily require special JSR-045 support in the IDE debugger implementation. SMAP allows you to set a default stratum for a class - and if the debugger doesn't explicitely request a specific stratum it will return the source for this stratum, not the default source from the bytecode.

So normally, setting the default stratum to `Mixin` should be enough to have the debuggers show the correct source file. Unfortunately, this is not the case for IntelliJ IDEA: The reason for that is that source files are primarily resolved by the fully qualified class name (instead of the source file name or the source path). Of course, the class name will always point to the original source of the target class.

I can imagine that it won't work in Eclipse for the same reason. Nevertheless, I believe that I can patch IntelliJ IDEA's implementation from MinecraftDev to make it work properly. If I get it working, would you agree with adding the source map generation to the Mixin core even if the IDEs don't implement it properly by default?

**Comment by Mumfrey** [comment]
It's worth bearing in mind that the code on that branch was just what was committed after a considerable number of painful hours trying all sorts of different things, and I'm pretty sure I actually only set the stratum name to `Java` right near the end just to try and get literally any response out of the debugger to the presence of the SMAP. I tested with strata named by the mixin, just named `Mixin` amongst many other things. That code was never meant to see the light of day in all honesty.

> So normally, setting the default stratum to Mixin should be enough to have the debuggers show the correct source file. Unfortunately, this is not the case for IntelliJ IDEA: The reason for that is that source files are primarily resolved by the fully qualified class name (instead of the source file name or the source path). Of course, the class name will always point to the original source of the target class.

It's not the case for Eclipse either - naturally, I was testing on my end with Eclipse. Perhaps this is something you can fix in the plugin.

> I can imagine that it won't work in Eclipse for the same reason. Nevertheless, I believe that I can patch IntelliJ IDEA's implementation from MinecraftDev to make it work properly. If I get it working, would you agree with adding the source map generation to the Mixin core even if the IDEs don't implement it properly by default?

Yes, my original intention with the SMAP stuff is that it would be basically just a standard feature (much like `MixinMerged` which would decorate the class being generated. I only binned it because it was clear it wasn't going to work without extending the IDE, and I don't want to ever become Lombok. If you get it working in IntelliJ, then there's impetus for me to try and get it working on Eclipse as well.

**Comment by Minecrell** [comment]
> If you get it working in IntelliJ, then there's impetus for me to try and get it working on Eclipse as well.

I have implemented it in https://github.com/minecraft-dev/MinecraftDev/pull/115/commits/601eb05dc33acb236670972877a7f803b17fa81f. The implementation is very similar to my previous one except that I was able to remove the parser of the custom format since I can just query the values from the SMAP using JDI.

The way it works is that IntelliJ IDEA allows registering multiple `PositionManager`s. They get called with the [JDI `Location`](https://docs.oracle.com/javase/8/docs/jdk/api/jpda/jdi/com/sun/jdi/Location.html) and are asked to return the source file together with the line number. I check if the default stratum is set to `Mixin` since that will be only the case for the classes modified by Mixin.

Then I can simply call [`Location.sourcePath()`](https://docs.oracle.com/javase/8/docs/jdk/api/jpda/jdi/com/sun/jdi/Location.html#sourcePath--) and [`Location.lineNumber()`](https://docs.oracle.com/javase/8/docs/jdk/api/jpda/jdi/com/sun/jdi/Location.html#lineNumber--) to get the path to the Mixin source file and the line number. I'm still looking up the classes by their qualified name (from the source path) since that is the easiest way to find the source file reliably in the workspace.

The other code in `MixinPositionManager` is necessary to redirect all the breakpoints to the target class (otherwise they won't be triggered). Finally, JDI's [`ReferenceType.locationsOfLine`](https://docs.oracle.com/javase/8/docs/jdk/api/jpda/jdi/com/sun/jdi/ReferenceType.html#locationsOfLine-int-) will automatically handle applying the line offset to the breakpoints in Mixin source files.

I've pushed the few changes to Mixin to the `sourcemap-rebased` branch on your server:

- Changed the stratum to `Mixin`
- SMAP defines line numbers starting with 1, the current code was assuming they start with 0
- Disabled the custom "pseudo generated source file" that changes the `SourceFile` attribute of the target class. I don't think it is needed for SMAP to work and it would only break the links in displayed stack traces.

Except for some general cleanup of the SMAP generation code we should also consider that it is possible that the target class has more than 3000 lines already or that one of the Mixins has more than 1000 lines. Maybe we should check for the maximum line number in the classes and apply the offsets based on them?

If you like I can write up a draft for the SMAP implementation with the cleanup and the changes I mentioned above tomorrow and create a PR once it is done so we can discuss the remaining implementation details.

**Comment by Mumfrey** [comment]
> * Disabled the custom "pseudo generated source file" that changes the SourceFile attribute of the target class. I don't think it is needed for SMAP to work and it would only break the links in displayed stack traces.

I only added that because I wasn't sure whether having the "real" source file in the `SourceFile` attribute was defeating the use of the smap. It was all test code.

> Except for some general cleanup of the SMAP generation code we should also consider that it is possible that the target class has more than 3000 lines already or that one of the Mixins has more than 1000 lines. Maybe we should check for the maximum line number in the classes and apply the offsets based on them?

Again, just an artefact of the fact that that code was only ever to try and figure out if it worked in principle, fixed-sized strata weren't intended to be a real production thing. It needs to be as compact as possible though because if I recall correctly line numbers can only go up to 16384 (I think, might be misremembering that).

> If you like I can write up a draft for the SMAP implementation with the cleanup and the changes I mentioned above tomorrow and create a PR once it is done so we can discuss the remaining implementation details.

I will take a look at what you've done and look at making it production-ready and see whether anything else needs to be taken into consideration. I would prefer if it wasn't PR'd.

**Comment by Minecrell** [comment]
> It needs to be as compact as possible though because if I recall correctly line numbers can only go up to 16384 (I think, might be misremembering that).

I'm not aware of such a limitation (at least I didn't see anything regarding this in the class file specification or in the debugger code) but on the other hand I'm not an expert in Java bytecode things.

> I will take a look at what you've done and look at making it production-ready and see whether anything else needs to be taken into consideration. I would prefer if it wasn't PR'd.

OK, let me know when you have pushed the change to a branch or similar, then I can test them with my changes in MinecraftDev so we can be sure it is working properly.

**Comment by Mumfrey** [comment]
> I'm not aware of such a limitation (at least I didn't see anything regarding this in the class file specification or in the debugger code) but on the other hand I'm not an expert in Java bytecode things.

Line numbers are stored as `short` in the bytecode from what I remember, so it might be 32768, either way there's definitely a limit.

> OK, let me know when you have pushed the change to a branch, then I can test them with my changes in MinecraftDev so we can be sure it is working properly.

I will take a look at it tomorrow.

**Comment by gabizou** [comment]
@Mumfrey I believe line numbers are stored as unsigned shorts (seeing as you can't have negative line numbers) effectively limiting the line numbers to 65536. That being said, it's a good sign that there's a high enough limit where we shouldn't have to worry about loading up a 65k line length file or writing a class as such.

**Comment by Mumfrey** [comment]
The updated code is pushed to the private repo if you want to see what changes I made after you did. I've squashed the branch and pushed the cleaned-up code to `feature/sourcemap`. My Jenkins has also been [instructed to build it](http://jenkins.liteloader.com/view/Other/job/Mixin%20Branch/) so it will be available on Maven for testing.

The line numbers are now properly computed, with some tolerances built in for testings purposes.

The biggest revelation has been that **this just works out of the box in Eclipse now**. Not sure when it changed (obviously some time in the 18 months since I roughed this out originally) but basically in Eclipse NEON.2 it just works straight away. You can't set a breakpoint in a mixin (I guess the relationship is "one way", bytecode -> sourceode) but you can "step into" the mixin when debugging.

Setting this to acceptance testing, pending your confirmation that it works with your intellij plugin.

**Comment by Minecrell** [comment]
@Mumfrey 
Thank you for making the changes. It works fine with the changes for IntelliJ IDEA.

If they are using the JDI API correctly the problem with the breakpoints in Eclipse is possibly similar to a problem I had when writing the implementation for IntelliJ IDEA: Breakpoints in Mixins were working if I also had at least one breakpoint somewhere in target class.

This was happening because IntelliJ IDEA has enable the breakpoints as soon as the class is being loaded in the JVM. This is implemented by listening using a [`ClassPrepareRequest`](https://docs.oracle.com/javase/8/docs/jdk/api/jpda/jdi/com/sun/jdi/request/ClassPrepareRequest.html) in the JDI API. So when I set a breakpoint in a Mixin class it was waiting for the Mixin class to be loaded and never enabled the breakpoint for the target class. I solved this by creating a `ClassPrepareRequest` for each of the Mixin target classes whenever a breakpoint is set in a Mixin class.

The problem in Eclipse could be similar to this, but of course it could be also something completely different. Anyway, feel free to merge your changes - everything is working for my changes in IntelliJ IDEA.
